test/unit/org/apache/cassandra/locator/DynamicEndpointSnitchTest.java

```diff
@@ -60,57 +130,229 @@ private static EndpointsForRange full(InetAddressAndPort... endpoints)
 
         /// Tests CASSANDRA-6683 improvements
         // make the scores differ enough from the ideal order that we sort by score; under the old
         // dynamic snitch behavior (where we only compared neighbors), these wouldn't get sorted
-        setScores(dsnitch, 20, hosts, 10, 70, 20);
-        order = full(host1, host3, host2);
-        Util.assertRCEquals(order, dsnitch.sortedByProximity(self, full(host1, host2, host3)));
+        dsnitch.reset();
+        setScores(dsnitch, 20, allHosts, 10, 70, 20);
+        order = full(hosts[1], hosts[3], hosts[2]);
+        Util.assertRCEquals(order, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3])));
+
+        order = full(hosts[4], hosts[1], hosts[3], hosts[2]);
+        Util.assertRCEquals(order, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3], hosts[4])));
+
+        setScores(dsnitch, 40, allHosts, 10, 10, 11);
+        order = full(hosts[4], hosts[1], hosts[2], hosts[3]);
+        Util.assertRCEquals(order, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3], hosts[4])));
+    }
+
+    // CASSANDRA-14459 improvements to add latency probes instead of resets
+    @Test
+    public void testLatencyProbeNeeded()
+    {
+        InetAddressAndPort self = hosts[0];
+
+        // Four hosts, two are very latent, but all are requested for ranking
+        dsnitch.receiveTiming(hosts[1], 20, LatencyMeasurementType.READ);
+        dsnitch.receiveTiming(hosts[2], 10, LatencyMeasurementType.READ);
+        dsnitch.receiveTiming(hosts[3], 1000, LatencyMeasurementType.READ);
+        dsnitch.receiveTiming(hosts[4], 1000, LatencyMeasurementType.READ);
+        dsnitch.updateScores();
+
+        EndpointsForRange orderBefore = full(hosts[2], hosts[1], hosts[3], hosts[4]);
+
+        Map<InetAddressAndPort, DynamicEndpointSnitch.AnnotatedMeasurement> measurements = dsnitch.getMeasurementsWithPort();
+        // At this point we haven't ranked any of hosts so we should need no probes even
+        DynamicEndpointSnitch.calculateProbes(measurements, dsnitch.dynamicLatencyProbeInterval);
+        assertFalse(measurements.values().stream().anyMatch(m -> m.probeTimerMillis > 0));
+        assertTrue(measurements.values().stream().allMatch(m -> m.probeFuture == null));
+
+        // Two hosts continue receiving traffic but the last two are always ranked by the snitch
+        for (int i = 0; i < 10; i++)
+        {
+            dsnitch.receiveTiming(hosts[1], 20, LatencyMeasurementType.READ);
+            dsnitch.receiveTiming(hosts[2], 10, LatencyMeasurementType.READ);
+            Util.assertRCEquals(orderBefore, dsnitch.sortedByProximity(self, full(hosts[1], hosts[2], hosts[3], hosts[4])));
+        }
+        dsnitch.updateScores();
+
+        int probesBefore = (int) ((DynamicEndpointSnitch.MIN_PROBE_INTERVAL_MS / dsnitch.dynamicLatencyProbeInterval) + 1);
```

Comment:
Probes before what? (Rename?)

Reply:
Ack, renamed to `intervalsBeforeProbesStart` to try to make this clearer.